Building GraphRAG Python MCP tools (#368)
neo4j-graphacademy/courses: asciidoc/courses/genai-mcp-build-custom-tools-python/modules/2-database-features/lessons/4-context-object/lesson.adoc

The Context object is automatically injected into tools and resources that request it.
It provides access to:

* **Lifespan resources** - Database connections, configuration, etc.
* **Logging methods** - Send messages to the client at different log levels
* **Progress reporting** - Show progress for long-running operations
* **Resource reading** - Access other resources from within tools
* **Session information** - Request metadata and client capabilities

== Accessing the Context

To use the Context in a tool or resource, simply add a parameter with the `Context` type annotation:

[source,python]
----
from mcp.server.fastmcp import Context, FastMCP

mcp = FastMCP("Movies GraphRAG Server")

@mcp.tool()
async def my_tool(query: str, ctx: Context) -> str:
    """A tool that uses the context."""
    
    # The context is automatically injected by FastMCP
    # The parameter can have any name, but must be type-annotated
    
    return await process_query(query, ctx)
----

[TIP]
.Choose any parameter name
====
The context parameter can have any name (`ctx`, `context`, `c`, etc.) as long as it has the `Context` type annotation.
====

== A Complete Example

> Let's look at a simple example that demonstrates using the Context object to track progress while running multiple database queries:

Reviewer: I wouldnt describe this as simple. The complexity just went up a lot.

```suggestion
Let's look at an example that demonstrates using the Context object to track progress while running multiple database queries:
```
